### PR TITLE
Fix: OnRamp discovery

### PR DESCRIPTION
**Pull request by asoliman92**
* Use another chainSelector other than self to get the destChainConfig
* Fix `onRampDynamicConfig` type to match on-chain

Note: It was silently failing to get the `onRampDynamicConfig` so we were seeing only the symptoms when we were not able to read NativeToken as the flow is:
1. Get Router address
2. Get OnRamp Address
3. Get wrapped token address from Router (step 1)
4. Get FeeQuoter address from OnRamp (step 2)
5. Get native token using fee quoter and the wrapped token address.

### CODE REVIEW

File: pkg/reader/ccip.go [hunk 1]
```diff
@@ -1087,22 +1100,27 @@ func (r *ccipChainReader) getDestinationData(
...
+type onRampDynamicConfig struct {
+	FeeQuoter              []byte `json:"feeQuoter"`
+	ReentrancyGuardEntered bool   `json:"reentrancyGuardEntered"`
+	MessageInterceptor     []byte `json:"messageInterceptor"`
+	FeeAggregator          []byte `json:"feeAggregator"`
+	AllowListAdmin         []byte `json:"allowListAdmin"`
+}
+
+type getOnRampDynamicConfigResponse struct {
+	DynamicConfig onRampDynamicConfig `json:"dynamicConfig"`
```
Comment (makramkd) [inline pkg/reader/ccip.go:1112]: Some comments on why this is structured this way would be useful

File: pkg/reader/ccip.go [hunk 2]
```diff
@@ -1166,17 +1187,20 @@ func (r *ccipChainReader) getOnRampDestChainConfig(
...
 				primitives.Unconfirmed,
 				map[string]any{
-					"destChainSelector": chainSel,
+					// Any other chain other than self should work as a chain don't have self as a destination
+					// Here we use destChain as the chainSel is looping over all source chains, so all of them should have
+					// a config for the destChain
+					"destChainSelector": r.destChain,
 				},
 				&resp,
 			)
```
Comment (makramkd) [inline pkg/reader/ccip.go:1206]: Please fix the error message below which is incorrect

```golang
return fmt.Errorf("failed to get onramp dest chain config: %w", err)
```

File: pkg/reader/ccip.go [hunk 3]
```diff
@@ -1166,17 +1187,20 @@ func (r *ccipChainReader) getOnRampDestChainConfig(
...
 			err := r.contractReaders[chainSel].ExtendedGetLatestValue(
 				ctx,
 				consts.ContractNameOnRamp,
 				consts.MethodNameOnRampGetDestChainConfig,
 				primitives.Unconfirmed,
 				map[string]any{
-					"destChainSelector": chainSel,
+					// Any other chain other than self should work as a chain don't have self as a destination
+					// Here we use destChain as the chainSel is looping over all source chains, so all of them should have
+					// a config for the destChain
```
Comment (makramkd) [inline pkg/reader/ccip.go:1202]: Wait, this comment doesn't make sense to me. You're getting the dest chain config from each onramp, correct? Then we should always use `r.destChain`. I believe this was the fix I made here but elected not to merge: https://github.com/smartcontractkit/chainlink-ccip/pull/236/files#diff-a16bcafc2684d6be248d96e8fdf5fa77ec115f311c0c73d13f3f942ad0edb0c4R1123

**Review by mateusz-sekara — APPROVED** [review]

File: pkg/reader/ccip.go [hunk 4]
```diff
@@ -625,6 +625,17 @@ func (r *ccipChainReader) GetRMNRemoteConfig(
 	}, nil
 }
 
+func (r *ccipChainReader) isValidAddress(address []byte, contractName string, chain cciptypes.ChainSelector) bool {
+	if len(address) == 0 {
+		r.lggr.Errorw("address for contract is empty",
```
Comment (makramkd) [inline pkg/reader/ccip.go:630]: Don't use error, it will spam the logs with a big stack trace. Warn here makes more sense to me

**Review by makramkd — CHANGES_REQUESTED** [review]

File: pkg/reader/ccip.go (hunk 1 above)
Comment (asoliman92) [reply pkg/reader/ccip.go:1112]: I wrote it and deleted it locally 😁. re-adding them 👍

File: pkg/reader/ccip.go (hunk 4 above)
Comment (mateusz-sekara) [reply pkg/reader/ccip.go:630]: Hmm, I would say it depends. If this is something that should never happen or happens because of the wrong configuration then error/critical is easier to spot in the logs. If this is a regular check that can fail from time to time then definitely should not be `error`

File: pkg/reader/ccip.go (hunk 3 above)
Comment (asoliman92) [reply pkg/reader/ccip.go:1202]: > You're getting the dest chain config from each onramp

Yes

> Then we should always use r.destChain

All dest configs have same router address (or maybe test address). Why should it be always `r.destChain`?  In this specific line of code it needs to be `r.destChain` but in general any address other than self that has a destChainConfig should work. I'll rephrase it to reduce confusion.

File: pkg/reader/ccip.go (hunk 4 above)
Comment (asoliman92) [reply pkg/reader/ccip.go]: I agree with @mateusz-sekara here. If this error was present I could've figured out somethings earlier.

File: pkg/reader/ccip.go [hunk 5]
```diff
@@ -625,6 +625,17 @@ func (r *ccipChainReader) GetRMNRemoteConfig(
 	}, nil
 }
 
+func (r *ccipChainReader) isValidAddress(address []byte, contractName string, chain cciptypes.ChainSelector) bool {
```
Comment (winder) [inline pkg/reader/ccip.go:628]: nit: if you return rather than logging this might be helpful for other functions as well.
```suggestion
func isValidAddress(address []byte, contractName string, chain cciptypes.ChainSelector) (bool, error) {
```

File: pkg/reader/ccip.go [hunk 6]
```diff
@@ -625,6 +625,17 @@ func (r *ccipChainReader) GetRMNRemoteConfig(
...
+	if len(address) == 0 {
```
Comment (winder) [inline pkg/reader/ccip.go:629]: Check for zero addresses as well? The discovery processor has this helper:
```
// isZero returns true if data is nil or all zeros, otherwise returns false.
func isZero(data []byte) bool {
	for _, v := range data {
		if v != 0 {
			return false
		}
	}
	return true
}
```

File: pkg/reader/ccip.go [hunk 7]
```diff
@@ -1087,22 +1100,31 @@ func (r *ccipChainReader) getDestinationData(
...
+	FeeAggregator          []byte `json:"feeAggregator"`
+	AllowListAdmin         []byte `json:"allowListAdmin"`
+}
+
+// We're wrapping the onRampDynamicConfig this way to map to on-chain return type which is a named struct
+// https://github.com/smartcontractkit/chainlink/blob/12af1de88238e0e918177d6b5622070417f48adf/contracts/src/v0.8/ccip/onRamp/OnRamp.sol#L328
+//
+//nolint:lll
+type getOnRampDynamicConfigResponse struct {
+	DynamicConfig onRampDynamicConfig `json:"dynamicConfig"`
```
Comment (winder) [inline pkg/reader/ccip.go:1116]: Which one of these is the fix? If you were to remove the unused fields are the others still returned?

Comment (asoliman92) [reply pkg/reader/ccip.go:1116]: Main fix is the wrapper. I didn't try removing the fields again after adding the wrapper though.

File: pkg/reader/ccip.go (hunk 6 above)
Comment (asoliman92) [reply pkg/reader/ccip.go:629]: I don't remember seeing the processor warnings about the empty router though while debugging 🤔. hmmmmm..... maybe I just missed it in the huge log file. Let me remove it so we don't have duplicates for now and keep my eye on it for next couple of PRs.

File: pkg/reader/ccip.go (hunk 5 above)
Comment (asoliman92) [reply pkg/reader/ccip.go]: I removed it completely for now.

**Review by makramkd — APPROVED** [review]

File: pkg/reader/ccip.go (hunk 7 above)
Comment (winder) [reply pkg/reader/ccip.go:1103]: I meant these fields:
```
type onRampDynamicConfig struct {
	FeeQuoter              []byte `json:"feeQuoter"`
-	ReentrancyGuardEntered bool   `json:"reentrancyGuardEntered"`
-	MessageInterceptor     []byte `json:"messageInterceptor"`
-	FeeAggregator          []byte `json:"feeAggregator"`
-	AllowListAdmin         []byte `json:"allowListAdmin"`
}
```

Hopefully removing those would be OK otherwise we have to worry about things breaking every time the onchain types change.

It probably works for ignoring the implicit structure for named return values also.

The default json encoding ignores "extra" data: https://play.golang.com/p/RN_ICGCaVqV

**Review by winder — APPROVED** [review]
LGTM

File: pkg/reader/ccip.go (hunk 7 above)
Comment (makramkd) [reply pkg/reader/ccip.go:1103]: Yeah, I was referring to those fields as well, I think we tried without them and it didn't work.
